rust_src/src/syntax.rs

```diff
@@ -104,4 +109,88 @@ pub fn copy_syntax_table(mut table: LispObject) -> LispObject {
     copy
 }
 
+/// Move point forward ARG words (backward if ARG is negative).
+/// If ARG is omitted or nil, move point forward one word.
+/// Normally returns t.
+/// If an edge of the buffer or a field boundary is reached, point is
+/// left there and the function returns nil.  Field boundaries are not
+/// noticed if `inhibit-field-text-motion' is non-nil.
+///
+/// The word boundaries are normally determined by the buffer's syntax
+/// table, but `find-word-boundary-function-table', such as set up
+/// by `subword-mode', can change that.  If a Lisp program needs to
+/// move by words determined strictly by the syntax table, it should
+/// use `forward-word-strictly' instead.
+#[lisp_fn(min = "0", intspec = "^p")]
+pub fn forward_word(arg: Option<EmacsInt>) -> bool {
+    let arg = arg.unwrap_or(1);
+    let cur_buf = ThreadState::current_buffer();
+    let point = cur_buf.pt;
+
+    let val = match unsafe { scan_words(point, arg) } {
+        0 => {
+            if arg > 0 {
+                cur_buf.zv
+            } else {
+                cur_buf.begv
+            }
+        }
+        n => n,
+    };
+
+    // Avoid jumping out of an input field.
+    let tmp = constrain_to_field(
+        Some(LispNumber::Fixnum(val as EmacsInt)),
+        LispNumber::Fixnum(point as EmacsInt),
+        false,
+        false,
+        Qnil,
+    ) as isize;
+    unsafe { set_point(tmp) };
+    val == tmp
```

Comment:
You are comparing `tmp` to `val` while the C code compared it to output of `scan_words`.